onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorQAttention.cpp

```diff
@@ -93,46 +123,36 @@ class DmlOperatorQAttention : public DmlOperator
         auto weightTensorShape = m_inputTensorDescs[weightsIndex].GetSizes();
         ML_CHECK_VALID_ARGUMENT(weightTensorShape.size() == 2);
         ML_CHECK_VALID_ARGUMENT(weightTensorShape[0] == inputTensorShape[2]);
+        ML_CHECK_VALID_ARGUMENT(weightTensorShape[1] % 3 == 0);
 
-        const auto qkvHiddenSizes = kernelCreationContext.GetOptionalAttributeVectorInt32(AttrName::QkvHiddenSizes);
         if (hasBias)
         {
             auto biasTensorShape = m_inputTensorDescs[biasIndex].GetSizes();
             ML_CHECK_VALID_ARGUMENT(biasTensorShape.size() == 1);
+            ML_CHECK_VALID_ARGUMENT(biasTensorShape[0] % 3 == 0);
             ML_CHECK_VALID_ARGUMENT(weightTensorShape[1] == biasTensorShape[0]);
-
-            if (qkvHiddenSizes.empty())
-            {
-                ML_CHECK_VALID_ARGUMENT(biasTensorShape[0] % 3 == 0);
-            }
         }
 
-        if (!qkvHiddenSizes.empty())
+        if (hasPast)
         {
-            ML_CHECK_VALID_ARGUMENT(qkvHiddenSizes.size() == 3);
-            ML_CHECK_VALID_ARGUMENT(qkvHiddenSizes[0] == qkvHiddenSizes[1]);
-        }
-        else
-        {
-            ML_CHECK_VALID_ARGUMENT(weightTensorShape[1] % 3 == 0);
+            ML_CHECK_VALID_ARGUMENT(m_outputTensorDescs.size() == 2);
```

Comment:
I can replace with `kernelCreationContext.IsOutputValid(presentIndex)`

But will there be a case where output is not present and m_outputTensorDescs contains empty TensorDesc ?

Reply:
Yes. Because we are not passing the output indices explicitly while calling the `DmlOperator::Initialize`, it will use `MLOperatorKernelCreationContext::GetOutputCount` which returns the max number of output for any operator.

[https://github.com/microsoft/onnxruntime/blob/efa0cc2562c28e6376717b46ebc83dd29b68d348/onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperator.cpp#L292](https://github.com/microsoft/onnxruntime/blob/efa0cc2562c28e6376717b46ebc83dd29b68d348/onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperator.cpp#L292)

Reply:
I see, did not know about that, Thank you!